pep-0572.rst

```diff
@@ -275,16 +275,47 @@ i`` part establishes that ``i`` is local to the comprehension, but the
 ``i :=`` part insists that ``i`` is not local to the comprehension.
 The same reason makes these examples invalid too::
 
-    [[(j := j) for i in range(5)] for j in range(5)]
-    [i := 0 for i, j in stuff]
-    [i+1 for i in i := stuff]
+    [[(j := j) for i in range(5)] for j in range(5)] # compiler TargetScopeError
+    [i := 0 for i, j in stuff]                       # compiler TargetScopeError
+    [i+1 for i in (i := stuff)]                      # compiler TargetScopeError
+
+While it's technically possible to assign consistent semantics to these cases,
+it's difficult to determine whether those semantics actually make *sense* in the
+absence of real use cases. Accordingly, the reference implementation will ensure
+that such cases raise ``TargetScopeError``, a new subclass of ``SyntaxError``
+(with the same signature), rather than executing with implementation defined
+behaviour.
+
+This restriction applies even if the assignment expression is never executed::
+
+    [False and (i := 0) for i, j in stuff]     # compiler TargetScopeError
+    [i for i, j in stuff if True or (j := 1)]  # compiler TargetScopeError
+
+For the comprehension body (the part before the first "for" keyword) and the
+filter expression (the part after "if" and before any nested "for"), this
+restriction applies solely to target names that are also used as iteration
+variables in the comprehension. Lambda expressions appearing in these
+positions introduce a new explicit function scope, and hence may use assignment
+expressions with no additional restrictions.
+
+Due to design constraints in the reference implementation, named expressions
```

Comment:
What are those design constraints? Are you saying that it would be better to allow named expressions in this position as long as they were parenthesized, but the implementation is too complicated? IIRC the PEP goes through great lengths not to use "do whatever's easiest in CPython" as an argument -- wouldn't this be effectively changing that?

Reply:
The case we really want to rule out is the one where the assignment expression target is also used as an iteration variable in the comprehension (the scoping design in PEP 572 ended up being elegant enough that the other edge cases all end up with behaviour that we're happy to make universal across implementations).

When we're inside the comprehension scope, we can do that without too much trouble, as the name binding operations are all part of the same section of the tree traversal, and we can check the assignment targets and iteration variables don't collide with each other - enforcing the constraint just needed an additional flag on some of the name bindings to say "This is a comprehension iteration variable".

The outermost iterable expression isn't like that though: since it gets passed in as an argument, any named expressions (and comprehensions!) that it contains are part of a separate branch of the compile tree, and walking back up won't see the iteration variable bindings. At the same time, walking up from the comprehension body can't currently tell the difference between local variables created by assignment expressions in the outermost iterable expression and local variable assignments created by earlier statements in the parent scope.

I can see a few ways we potentially *could* teach the symbol table analyser to tell the difference, but they'd all require additional machinery that we don't currently have, whereas the simpler rule "Don't allow assignment expressions in comprehension iterable expressions" just requires keeping a counter to keep track of each time we enter and leave one.

Thus this revised proposal: implement the simpler constraint now, and revisit the more complex one when presented with a concrete use case that the simpler constraint disallows.

Reply:
I added some more details to the PEP itself. It's still a bit hand-wavey, but it's hard to explain the specifics without first explaining exactly how the symbol table analysis is structured.